File: typescript/src/config/mode.ts
// Resolves `DEEPEVAL_MODE`: `stable` (default) or `experimental`.
//
// `experimental` opts a user into the newest deepeval features before they
// are finalised; anything gated on it may change or break between releases.
// `stable` is the default and the fallback for an unset or unrecognised value,
// so a typo can never silently enrol someone.
//
// Gate features with `isExperimental()` rather than comparing strings.

import { DEEPEVAL_MODE } from "@/constants";
// Cycle (keystore -> schema -> mode) is safe: only used inside a function.
import { readKeystore } from "@/config/keystore";

export const MODE_STABLE = "stable";
export const MODE_EXPERIMENTAL = "experimental";
export type DeepEvalMode = typeof MODE_STABLE | typeof MODE_EXPERIMENTAL;

export const DEFAULT_MODE: DeepEvalMode = MODE_STABLE;
export const SUPPORTED_MODES: readonly DeepEvalMode[] = [
  MODE_STABLE,
  MODE_EXPERIMENTAL,
];

const warnedUnrecognised = new Set<string>();

/**
 * Normalize a raw `DEEPEVAL_MODE` value. Returns `undefined` for values that
 * are unset, blank or unrecognised.
 */
export function normalizeDeepEvalMode(
  value: string | undefined | null,
): DeepEvalMode | undefined {
  if (value === undefined || value === null) return undefined;
  const s = value.trim().toLowerCase();
  if (s === MODE_STABLE) return MODE_STABLE;
  if (s === MODE_EXPERIMENTAL) return MODE_EXPERIMENTAL;
  return undefined;
}

/**
 * `"stable"` (default) or `"experimental"`.
 *
 * Mirrors Python's `resolve_deepeval_mode()`: the raw value comes from
 * `process.env` first, then the `.deepeval` keystore (the CLI's `set-mode`
 * writes there when `--save` is omitted). An unrecognised value warns once and
 * falls back to `stable`.
 */
export function resolveDeepEvalMode(
  env: NodeJS.ProcessEnv = process.env,
): DeepEvalMode {
  let raw = env[DEEPEVAL_MODE];
  if (raw === undefined || raw.trim() === "") {
    raw = readKeystore()[DEEPEVAL_MODE];
  }
  const mode = normalizeDeepEvalMode(raw);
  if (mode !== undefined) return mode;
  if (raw !== undefined && raw.trim() !== "" && !warnedUnrecognised.has(raw)) {
    warnedUnrecognised.add(raw);
    console.warn(
      `Warning: unrecognised ${DEEPEVAL_MODE}=${JSON.stringify(raw)}; ` +
        `falling back to '${DEFAULT_MODE}'. Valid values: ${SUPPORTED_MODES.join(", ")}.`,
    );
  }
  return DEFAULT_MODE;
}

/** True when the user has opted into experimental deepeval features. */
export function isExperimental(): boolean {
  return resolveDeepEvalMode() === MODE_EXPERIMENTAL;
}

File: deepeval/config/mode.py
"""Resolves `DEEPEVAL_MODE`: `stable` (default) or `experimental`.

`experimental` opts a user into the newest deepeval features before they are
finalised; anything gated on it may change or break between releases.
`stable` is the default and the fallback for an unset or unrecognised value,
so a typo can never silently enrol someone.

Gate features with `is_experimental()` rather than comparing strings. The list
of current experimental features and where they live is in `EXPERIMENTAL.md`.
"""

import os
import sys
from enum import Enum
from typing import Optional

MODE_ENV_VAR = "DEEPEVAL_MODE"


class DeepEvalMode(str, Enum):
    STABLE = "stable"
    EXPERIMENTAL = "experimental"

    def __str__(self) -> str:
        return self.value


DEFAULT_MODE = DeepEvalMode.STABLE
SUPPORTED_MODES = tuple(m.value for m in DeepEvalMode)

# Shell-friendly spellings accepted by the CLI and env var.
_MODE_ALIASES = {
    "": None,
    "stable": DeepEvalMode.STABLE,
    "experimental": DeepEvalMode.EXPERIMENTAL,
}

_warned_unrecognised: set = set()


def normalize_deepeval_mode(value: Optional[str]) -> Optional[DeepEvalMode]:
    """Return the `DeepEvalMode` for a raw value, or `None` if it is unset,
    blank or unrecognised."""
    if value is None:
        return None
    return _MODE_ALIASES.get(str(value).strip().lower())


def resolve_deepeval_mode() -> DeepEvalMode:
    """`DeepEvalMode.STABLE` (default) or `DeepEvalMode.EXPERIMENTAL`.

    Reads the env var directly (the `.deepeval` keystore is merged into the
    environment when settings load) so the choice can be flipped per-process.
    Unrecognised values fall back to `stable` with a one-time stderr warning.
    """
    raw = os.getenv(MODE_ENV_VAR) or ""
    mode = normalize_deepeval_mode(raw)
    if mode is not None:
        return mode
    if raw.strip() and raw not in _warned_unrecognised:
        _warned_unrecognised.add(raw)
        print(
            f"Warning: unrecognised {MODE_ENV_VAR}={raw!r}; falling back to "
            f"'{DEFAULT_MODE}'. Valid values: {', '.join(SUPPORTED_MODES)}.",
            file=sys.stderr,
        )
    return DEFAULT_MODE


def is_experimental() -> bool:
    """True when the user has opted into experimental deepeval features."""
    return resolve_deepeval_mode() == DeepEvalMode.EXPERIMENTAL
